Item disassembling QoL changes

GoLoT opened the pull request:
#### Summary
SUMMARY: Interface "Item disassembling screen QoL changes"

#### Purpose of change
Closes #27358 - Disassemble menu: Filter by result

#### Describe the solution
Added a new flag to inventory presets that controls whether items should be stacked if their components differ. It is set to false by default and (currently) only set to true in the `disassemble_inventory_preset`. This flag is checked when adding items to the inventory set and passed to the `stacks_with()` and `restack_items()` functions.

`stacks_with()` and `restack_items()` have been modified accordingly to receive an extra optional boolean parameter (false by default so it doesn't break existing code) that handles (un)stacking items based on their components.

To list the components for an item it lists the item's components and if it doesn't have any (i.e. not a crafted item) then the default uncraft recipe is used. Those components are also updated on the disassembling prompt that previously wrongly listed the default recipe components.

Then a new filter category was added to inventory screens to allow filtering by components. The category key is `d` for (d)isassemble. As it is implemented as a global item filter it works on any inventory screen that can be filtered but on any screens that don't handle stacking items with different components it only searches for components for the first item in a stack. In a future PR the advanced inventory screen could be updated to re-stack items when the d search flag is used so items with a specific component can be found and transfered easily.

Also called the map redraw function after setting a new filter so the previous unfiltered window is removed before drawing the new filtered one, preventing the glitch where old windows show behind the new window when it is resized to a smaller size. If the old window overlaps the sidebar it still shows that part of the old window so a full UI redraw might be needed instead. Hopefully this won't break the behavior of any other inventory windows.

EDIT: As per @jbytheway 's suggestion: Added a new function `item::get_uncraft_components()` that returns a `requirement_data::alter_item_comp_vector` with the components used to craft the item or the default recipe components if it wasn't crafted.

#### Describe alternatives you've considered
None.

#### Additional context
![Imgur](https://i.imgur.com/0YyZXCj.png)
![Imgur](https://i.imgur.com/KaIMYuQ.png)
![Imgur](https://i.imgur.com/2YATqRw.png)
![Imgur](https://i.imgur.com/5vQz9qL.png)

GoLoT commented:
jenkins rebuild

NotFuji commented:
What does the UI look like when items have a large number of components, like robot corpses?

GoLoT commented:
It looks the same way it does before the changes, the only thing the PR changes is the way items are stacked in the disassemble inventory so items crafted with different components are shown in separate stacks. The windows is stretched to fit all components and if it fills the screen then it switches to a single-column layout so it doesn't overflow.

Maybe a new change can be suggested to make multi-line entries in the menu, but that would affect all inventory windows in the game, I guess. Better to leave that for a new PR.

jbytheway commented on src/item.cpp:
```diff
@@ -600,6 +600,39 @@ bool item::stacks_with( const item &rhs ) const
+                                       rhs.components[i].count();
+                if( lhs_type != rhs_type || lhs_count != rhs_count ) {
+                    return false;
+                }
```
I would hope that this whole section could be simplified by calling the hypothetical new function I suggested above to get the list of components for each, then just comparing those.

jbytheway commented on src/item.cpp:
```diff
@@ -600,6 +600,39 @@ bool item::stacks_with( const item &rhs ) const
+            }
+        }
+    }
     return std::equal( contents.begin(), contents.end(), rhs.contents.begin(), []( const item & a,
```
This call to `std::equal` logically belongs with the size comparison above.  They were adjacent before you inserted the new condition.  They should remain adjacent to maximize readability.

jbytheway commented on src/item_search.cpp:
```diff
@@ -43,6 +43,31 @@ std::function<bool( const item & )> basic_item_filter( std::string filter )
+                    //If item wasn't crafted use default recipe components
+                    const auto &req = recipe_dictionary::get_uncraft( i.typeId() ).disassembly_requirements();
+                    if( req.is_empty() ) {
+                        return false;
```
Is this early exit really necessary?  Is `req.get_components()` incorrect or expensive to call when `req.is_empty()`?

GoLoT replied:
Removed.

jbytheway commented on src/item_search.cpp:
```diff
@@ -43,6 +43,31 @@ std::function<bool( const item & )> basic_item_filter( std::string filter )
+                        }
+                    }
+                    return false;
+                }
```
Again this is duplicating the component extraction logic.

GoLoT replied:
Changed it to use the proposed auxiliary function.

jbytheway commented on src/output.cpp:
```diff
@@ -761,9 +761,9 @@ void draw_item_filter_rules( const catacurses::window &win, int starty, int heig
+                              _( "Search [c]ategory, [m]aterial, [q]uality or [d]isassembled components:" ) );
     //~ An example of how to filter items based on category or material.
-    fold_and_print( win, starty, 1, len, c_white, _( "Example: c:food,m:iron,q:hammering" ) );
+    fold_and_print( win, starty, 1, len, c_white, _( "Example: c:food,m:iron,q:hammering,d:pipe" ) );
```
Since this line is changing anyway, I'd say this should be "Examples:", not "Example:".

GoLoT replied:
Done.

GoLoT commented on src/game_inventory.cpp:
```diff
@@ -306,16 +306,26 @@ class disassemble_inventory_preset : public pickup_inventory_preset
+                    []( const decltype( i->components )::value_type & comps ) {
+                        return item_comp( comps.typeId(), comps.count() ).to_string();
+                    } );
+                }
```
Added a new function `item::get_uncraft_components()` that returns a `requirement_data::alter_item_comp_vector` with the components used to craft the item or the default recipe components if it wasn't crafted.

GoLoT commented:
Thanks @jbytheway for the review. Implemented all suggested changes.

jbytheway commented on src/item.cpp:
```diff
@@ -7330,3 +7346,17 @@ int item::get_min_str() const
     }
 }
+
+requirement_data::alter_item_comp_vector item::get_uncraft_components() const
```
I think a better alternative here is to return a `std::vector<item_comp>` rather than `requirement_data::alter_item_comp_vector`.  When `components.empty()` you have to extract the first element of each list of alternatives, rather than just return the recipe requirements verbatim.  That might seem like extra work, but actually it's extra code but less work because the current implementation has to copy the whole vector of vectors, which is probably more work for the program.

And overall it should be a code-size win, because the new code in the above implementation of `stacks_with` can be much simpler: just test `get_uncraft_components() != rhs.get_uncraft_components()`.  The other uses of `get_uncraft_components` should get slightly simpler too, but not such a big difference.

GoLoT replied:
Changed the auxiliary function to return a `std::vector<item_comp>` and modified calling code accordingly. I had to add equality operators to the `item_comp` struct for the vector comparison to work but it made the code more readable as you said.

About the `player:disassemble` thing, it only uses the recipe to check if the item can be disassembled. I left it that way because the new function only returns a list of materials without any information about the recipe so the recipe lookup is still needed. The actual disassembly is handled afterwards as an action and it uses the components in the item, the only thing that was wrong was the item list that was shown in the disassemble query, which is fixed already by using the new function to get the components list.

GoLoT commented:
Added the WIP tag because after doing some testing I realized that handling stacks in player inventory is messy and doesn't work as expected. The current solution creates a copy of the list of stacks, restacks them and uses that list to build the item list. For items on the ground it works just fine but accessing a specific item in a stack in the player inventory isn't as straightforward, all split stacks point to the first item in the original stack.

I have a solution in mind that involves restacking the player inventory (separating items by their components) so the listed stacks match the stacks in the inventory and restacking again after disassembly. I will try to find an alternative that allows using a temporary list without messing with the inventory but this seems to be the easiest solution.

GoLoT commented:
Ok, so the problem wasn't with the pre-stacking of items in player inventory but with how disassemble handles retrieving the items. The custom stacks work just fine but the item was being removed from its current location and added to the player inventory forcing it to be stacked again without checking for components compatibility. This caused the items to sometimes be stacked with existing stacks in the player inventory that didn't match their components.

I added a flag to the `item_location` class that controls if functions like `obtain()` should stack the items with existing ones or make a new stack for that item. This way items obtained for disassembly can be placed into their own stack and be accessed independently through the `item_location` object. I'm not sure if adding the flag to `item_location` is the best way to do it but I can't think of a better solution. I'm open to suggestions.

Anyway, I removed the WIP as it seems to be working fine now in all my tests cases.

GoLoT commented:
The problem wasn't what I thought at first. The implementation was fine but had its limitations for my use case, mainly that there wasn't a way to move an item to an inventory without it stacking with other items and I expected the obtain functions to return a location to a specific item rather than a whole stack. The combination of the two was the actual problem. Adding the new `should_stack` flag takes care of it and can be used elsewhere if needed.

I don't think there is a need for a unit test but maybe it's worth rewriting some of the inventory implementation to avoid problems like this in the future in addition to combining the flags as you mentioned. Being able to locate items inside a stack with an `item_location` would solve many issues I think because right now it only points to the stack. When I have time I can take a look at it and see what can be done.
